File: 125.验证回文串.go
/*
 * @lc app=leetcode.cn id=125 lang=golang
 *
 * [125] 验证回文串
 */

// @lc code=start
func isPalindrome(s string) bool {
	left, right := 0, len(s) - 1
	for left < right {
		numStatus := false
		// 越界判定要在前面
		for left < right && !isValid(s[left]) {
			// 不为有效字符左指针前移
			left++
		}
		for left < right && !isValid(s[right]) {
			// 同上,右指针右移
			right--
		}
		if isNumeric(s[left]) || isNumeric(s[right]) {
			// 存在数字
			numStatus = true
		}
		// 字符相同,字母大小写判定，大小写不能出现数字
		if s[left] == s[right] || (!numStatus && (int(s[left]) + 32 == int(s[right]) || int(s[right]) + 32 == int(s[left]))) {
			left++
			right--
		} else {
			return false
		}
	}
	return true
}

func isValid(s byte) bool {
	// 有效字符为数字,大小写英文字母
	if (s >= '0' && s <= '9') || (s >= 'a' && s <= 'z') || (s >= 'A' && s <= 'Z') {
		return true
	}
	return false
}

func isNumeric(s byte) bool {
	if s >= '0' && s <= '9' {
		return true
	}
	return false
}
// @lc code=end

File: 125.验证回文串.py
#
# @lc app=leetcode.cn id=125 lang=python3
#
# [125] 验证回文串
#

# @lc code=start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1

        def validPart(one):     # 字符类型判定
            if one.isdigit() or ord("a") <= ord(one) <= ord("z") or ord("A") <= ord(one) <= ord("Z"):
                return True
            return False

        while left < right:
            num_status = False
            while left < right and not validPart(s[left]):
                left += 1
            while left < right and not validPart(s[right]):
                right -= 1
            if s[left].isdigit() or s[right].isdigit():     # 存在数字
                num_status = True
            if s[left] == s[right] or ((ord(s[left]) + 32 == ord(s[right]) or ord(s[right]) + 32 == ord(s[left])) and not num_status):     # 字符相同继续
                left += 1
                right -= 1
            else:   # 不相同退出
                return False
        # 无不同
        return True

# @lc code=end
